(Feature) HD wallets' support

vbaranov opened the pull request:
Menu item:
![screen shot 2018-11-08 at 14 13 09](https://user-images.githubusercontent.com/4341812/48195641-442e0680-e361-11e8-89be-9014a134c781.png)


Choose HD wallet page:
![screen shot 2018-11-08 at 14 13 42](https://user-images.githubusercontent.com/4341812/48195640-43957000-e361-11e8-98a1-dcae5768c6cb.png)


Trezor:
![screen shot 2018-11-08 at 14 17 19](https://user-images.githubusercontent.com/4341812/48195639-43957000-e361-11e8-9959-f3efca7aaf32.png)


Ledger:
![screen shot 2018-11-08 at 14 18 28](https://user-images.githubusercontent.com/4341812/48195638-43957000-e361-11e8-817b-7cadf81c83ec.png)



Guide to configure Trezor simulator: https://github.com/poanetwork/nifty-wallet/blob/master/docs/trezor-emulator.md

dennis00010011b commented:
@vbaranov 
Improper error message if 2nd attempt to connect to Trezor

<img width="1388" alt="screen shot 2018-11-08 at 5 09 33 pm" src="https://user-images.githubusercontent.com/14067096/48237082-f70a5e80-e379-11e8-805a-b13da65d5fc8.png">

Steps (don't plug-in HD wallet)
  1. Go to  'Connect to hardware wallet'
 2. Select 'Trezor', click 'Connect'
 3. Switch account or refresh the page, main screen will opens
 4.  Go to  'Connect to hardware wallet'
 5. Select 'Trezor', click 'Connect'

dennis00010011b commented:
@vbaranov 
Misspelling in notification. Letter 'd' in word 'device' should be capital
<img width="363" alt="screen shot 2018-11-12 at 5 45 37 pm" src="https://user-images.githubusercontent.com/14067096/48386634-7f08a500-e6a7-11e8-974e-09277aa7ae79.png">

dennis00010011b commented:
@vbaranov 
App is frozen if attempting to delete Trezor's account after choice to forget device 

Steps for reproduce:
1. Connect TrezorOne device to PC
2. Open NW, import any account from device
3. Go to Accounts-> Connect hw  and click 'Forget this device'
4. Open Accounts and select close icon for deleting Trezor's account

Actual result:
 - app unresponsive
 - uncaught errors in console [fmdpnlapjechkkicpkahdmanomngknlf-1542081145709.txt](https://github.com/poanetwork/nifty-wallet/files/2574606/fmdpnlapjechkkicpkahdmanomngknlf-1542081145709.txt)

Expected result: 
 - user should be able to delete hw account at any time 
OR
- hw accounts should be removed if user selects to forget device

<img width="1440" alt="screen shot 2018-11-12 at 7 52 14 pm" src="https://user-images.githubusercontent.com/14067096/48390449-f1cd4c80-e6b6-11e8-9bb7-8f3b1762f537.png">

vbaranov commented:
@dennis00010011b 

>Improper error message if 2nd attempt to connect to Trezor

Fixed here https://github.com/poanetwork/nifty-wallet/pull/183/commits/a33ba78ff1f970ba52c6e75f3ae802e89d665101

>App is frozen if attempting to delete Trezor's account after choice to forget device

Good catch! Also, fixed here https://github.com/poanetwork/nifty-wallet/pull/183/commits/a33ba78ff1f970ba52c6e75f3ae802e89d665101

>Misspelling in notification. Letter 'd' in word 'device' should be capital

Fixed  here https://github.com/poanetwork/nifty-wallet/pull/183/commits/bad4943a21dc4a2f657d8acef4b247f7df398aea

dennis00010011b commented:
> > Improper error message if 2nd attempt to connect to Trezor
> 
> Fixed here [a33ba78](https://github.com/poanetwork/nifty-wallet/commit/a33ba78ff1f970ba52c6e75f3ae802e89d665101)
> 
> > App is frozen if attempting to delete Trezor's account after choice to forget device
> 
> Good catch! Also, fixed here [a33ba78](https://github.com/poanetwork/nifty-wallet/commit/a33ba78ff1f970ba52c6e75f3ae802e89d665101)
> 
> > Misspelling in notification. Letter 'd' in word 'device' should be capital
> 
> Fixed here [bad4943](https://github.com/poanetwork/nifty-wallet/commit/bad4943a21dc4a2f657d8acef4b247f7df398aea)

Tested it with Trezor One device. No issues were found

dennis00010011b commented:
@vbaranov 
Uncaught error in console when try to export private key for Trezor's account
<img width="1217" alt="screen shot 2018-11-13 at 12 14 09 pm" src="https://user-images.githubusercontent.com/14067096/48440477-dd806280-e73d-11e8-876b-276a41944fa4.png">

dennis00010011b commented:
@vbaranov 
Ledger: notification's text is incorrect when a transaction failed
<img width="374" alt="screen shot 2018-11-13 at 2 23 23 pm" src="https://user-images.githubusercontent.com/14067096/48447587-31e10d80-e751-11e8-8dcd-9d50d2e02015.png">

dennis00010011b commented:
@vbaranov 
 Error messages should be more friendly and understandable for non-proficient user. Something like that:

1. Please plug in Ledger and try again
<img width="367" alt="screen shot 2018-11-13 at 4 22 07 pm" src="https://user-images.githubusercontent.com/14067096/48452369-aa50ca00-e763-11e8-93fb-05493eee664f.png">

2. Can not recognize device

<img width="371" alt="screen shot 2018-11-13 at 4 30 08 pm" src="https://user-images.githubusercontent.com/14067096/48452376-b046ab00-e763-11e8-8772-06dfc4495c11.png">

dennis00010011b commented:
@vbaranov 
Ledger: list of accounts is lost (reconnection is needed ) when user exports the  private key for any account
https://www.useloom.com/share/b83694df7c544a59b2849b33d9ac4899
Steps:
 1. Connect Ledger, unlock any account
 2. Export PR for any account
 3. Go to Accounts-> Connect HW

 Expected result: list of accounts should be displayed
 Actual result: reconnection is needed

Same things happens when delete account
Device isn't really disconnected, just account list isn't displayed

vbaranov commented:
@dennis00010011b 

>Ledger: notification's text is incorrect when a transaction rejected by device

This has been fixed.

Please, create separate issues for other findings. I think we could solve them in a separate PRs. They require some sort of refactoring. As about friendly errors, of course, we can parse errors from HD wallet and display friendly errors, but it will be hard to manage because errors definitions from HD device are out from our control, and if errors will be changed on HD wallet side, every time we will need to change parsing on NW side.

fvictorio commented on old-ui/app/components/connect-hardware/account-list.js:
```diff
@@ -0,0 +1,194 @@
+      )
+    }
+
+    capitalizeDevice = (device) => {
```
This function seems to do the same as `capitalizeFirstLetter`.

vbaranov replied:
Thanks. Duplicate has been removed https://github.com/poanetwork/nifty-wallet/pull/183/commits/256e58527485a7622f8049d2c03f53aca058b994
